service/storage_service.cc

```diff
@@ -940,39 +940,61 @@ future<> storage_service::handle_state_normal(inet_address endpoint) {
     // the collision is resolved by assigning the token to the endpoint which started later.
     std::unordered_set<token> owned_tokens;
 
+    // token_to_endpoint_map is used to track the current token owners.
+    // when any tokne is replaced by a new owner, we track the existing owner in `candidates_for_removal`
+    // and eventually, if any candidate for removal end up owning no tokens, it is removed
+    // from token_metadata.
+    std::unordered_map<token, inet_address> token_to_endpoint_map = get_token_metadata().get_token_to_endpoint();
+    std::unordered_set<inet_address> candidates_for_removal;
+
     for (auto t : tokens) {
         // we don't want to update if this node is responsible for the token and it has a later startup time than endpoint.
-        auto current_owner = tmptr->get_endpoint(t);
-        if (!current_owner) {
+        auto current = token_to_endpoint_map.find(t);
+        if (current == token_to_endpoint_map.end()) {
             slogger.debug("handle_state_normal: New node {} at token {}", endpoint, t);
             owned_tokens.insert(t);
-        } else if (endpoint == *current_owner) {
-            slogger.debug("handle_state_normal: endpoint={} == current_owner={} token {}", endpoint, *current_owner, t);
+            continue;
+        }
+        auto current_owner = current->second;
+        if (endpoint == current_owner) {
+            slogger.debug("handle_state_normal: endpoint={} == current_owner={} token {}", endpoint, current_owner, t);
             // set state back to normal, since the node may have tried to leave, but failed and is now back up
             owned_tokens.insert(t);
-        } else if (_gossiper.compare_endpoint_startup(endpoint, *current_owner) > 0) {
-            slogger.debug("handle_state_normal: endpoint={} > current_owner={}, token {}", endpoint, *current_owner, t);
+            token_to_endpoint_map.erase(current);
+        } else if (_gossiper.compare_endpoint_startup(endpoint, current_owner) > 0) {
+            slogger.debug("handle_state_normal: endpoint={} > current_owner={}, token {}", endpoint, current_owner, t);
             owned_tokens.insert(t);
+            slogger.info("handle_state_normal: remove endpoint={} token={}", current_owner, t);
```

Comment:
If I understand correctly...

This code was practically dead before. The most common case of stealing tokens was node replace operation. But in this case the replacing node also stole the Host ID. Then we would call `tmptr->remove_endpoint` for the replaced node (in `do_remove_node`) earlier above, so by the time we entered this `for (auto t : tokens)` loop, the replaced node was no longer a member of `tmptr`. We then called `tmptr->get_endpoint` which returned `nullopt`, and we logged:
```
            slogger.debug("handle_state_normal: New node {} at token {}", endpoint, t);
```

The other case was token collision. But token collisions were (are) extremely rare. And even though this code was incorrect, it didn't matter: for the incorrectness to actually make a difference, two nodes would have to collide on *every* token, which is practically impossible given 256 tokens. Even with 1 common token it's extremely unlikely (P=2^(-64)). That's why the bug lived so long unnoticed.

Anyway, what will happen now: we're no longer using `tmptr`, we're using a fresh copy of token metadata obtained with `get_token_metadata()`. So during the common case of replace operation, we will log the fact of stealing every token on INFO level, producing a bit of spam in the logs. Even if we continued using `tmptr`, then combined with https://github.com/scylladb/scylladb/pull/12084, where we no longer steal the Host ID, `tmptr` still does have the replaced endpoint inside this loop, so we would again rely on this code and produce the log spam.

Point being: I think we should combine these separate log calls, one for every token, to a single log call that will print the list of all stolen tokens, so we don't spam the logs with 256 messages on every node.